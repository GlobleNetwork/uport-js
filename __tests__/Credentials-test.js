import Credentials from '../src/Credentials'
import { createJWT } from '../src/JWT'
import SimpleSigner from '../src/SimpleSigner'
import { SECP256K1Client, TokenVerifier, decodeToken } from 'jsontokens'
import nock from 'nock'
import MockDate from 'mockdate'
MockDate.set(1485321133 * 1000)

const privateKey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const publicKey = SECP256K1Client.derivePublicKey(privateKey)
const signer = SimpleSigner(privateKey)
const verifier = new TokenVerifier('ES256K', publicKey)
const profileA = {publicKey, name: 'David Chaum'}
const registry = (address) => new Promise((resolve, reject) => resolve(address === '0x001122' ? profileA : null))
const uport = new Credentials({signer, address: '0x001122', registry})
const uport2 = new Credentials({registry})

describe('createRequest', () => {
  it('creates a valid JWT for a request', () => {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return expect(verifier.verify(jwt)).toBeTruthy()
    })
  })

  it('has correct payload in JWT for a plain request for public details', () => {
    return uport.createRequest().then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT requesting a specific network_id', () => {
    return uport.createRequest({network_id: '0x4'}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('ignores unsupported request parameters', () => {
    return uport.createRequest({signing: true, sellSoul: true}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request', () => {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request asking for verified credentials', () => {
    return uport.createRequest({requested: ['name', 'phone'], verified: ['name']}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request with callbackUrl', () => {
    return uport.createRequest({ requested: ['name', 'phone'], callbackUrl: 'https://myserver.com'}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request for push notifications', () => {
    return uport.createRequest({ requested: ['name', 'phone'], notifications: true }).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })
})

describe('attest', () => {
  it('creates a valid JWT for an attestation', () => {
    return uport.attest({sub: '0x112233', claim: {email: 'bingbangbung@email.com'}, exp: 1485321133 + 1}).then((jwt) => {
      return expect(verifier.verify(jwt)).toBeTruthy()
    })
  })

  it('has correct payload in JWT for an attestation', () => {
    return uport.attest({sub: '0x112233', claim: {email: 'bingbangbung@email.com'}, exp: 1485321133 + 1}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })
})

describe('receive', () => {

  function createShareResp (payload = {}) {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return createJWT({address: '0x001122', signer}, {...payload, type: 'shareResp', req:jwt})
    })
  }

  function createShareRespMissingRequest (payload = {}) {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return createJWT({address: '0x001122', signer}, {...payload, type: 'shareResp'})
    })
  }

  function createShareRespWithExpiredRequest (payload = {}) {
    return uport.createRequest({requested: ['name', 'phone'], exp: Date.now() - 1}).then((jwt) => {
      return createJWT({address: '0x001122', signer}, {...payload, type: 'shareResp', req:jwt})
    })
  }

  function createShareRespWithVerifiedCredential (payload = {}, verifiedClaim = {sub: '0x112233', claim: {email: 'bingbangbung@email.com'}, exp: 1485321133 + 1}) {
    return uport.attest(verifiedClaim).then(jwt => {
      return createShareResp({...payload, verified: [jwt]})
    })
  }

  it('returns profile mixing public and private claims', () => {
    return createShareResp({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile mixing public and private claims and verified credentials', () => {
    return createShareRespWithVerifiedCredential({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile with only public claims', () => {
    return createShareResp().then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile with private chain network id claims', () => {
    return createShareResp({nad: '34wjsxwvduano7NFC8ujNJnFjbacgYeWA8m'}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns pushToken if available', () => {
    return createShareResp({capabilities: ['PUSHTOKEN']}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile.pushToken).toEqual('PUSHTOKEN')
    )
  })

  it('handles response to expired request', () => {
    return createShareRespWithExpiredRequest().then(jwt => uport.receive(jwt)).catch(error => expect(error.message).toEqual('JWT has expired'))
  })

  it('handles response with missing challenge', () => {
    return createShareRespMissingRequest().then(jwt => uport.receive(jwt)).catch(error => expect(error.message).toEqual('Challenge was not included in response'))
  })

/////////////////////////////// no address in uport settings ///////////////////////////////

  it('returns profile mixing public and private claims', () => {
    return createShareResp({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport2.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile mixing public and private claims and verified credentials', () => {
    return createShareRespWithVerifiedCredential({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport2.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile with only public claims', () => {
    return createShareResp().then(jwt => uport2.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile with private chain network id claims', () => {
    return createShareResp({nad: '34wjsxwvduano7NFC8ujNJnFjbacgYeWA8m'}).then(jwt => uport2.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns pushToken if available', () => {
    return createShareResp({capabilities: ['PUSHTOKEN']}).then(jwt => uport2.receive(jwt)).then(profile =>
      expect(profile.pushToken).toEqual('PUSHTOKEN')
    )
  })

/////////////////////////////// no address in uport settings///////////////////////////////
})

describe('push', () => {
  const PUSHTOKEN = 'SECRETPUSHTOKEN'
  const payloadUrl = 'me.uport:me'

  beforeEach(() => {
    nock.disableNetConnect()
  })

  afterEach(() => {
    nock.enableNetConnect()
  })

  it('pushes url to pututu', () => {
    nock('https://pututu.uport.me', {
      reqheaders: {
        'authorization': `Bearer ${PUSHTOKEN}`
      }
    })
    .post('/api/v1/sns', {url: payloadUrl})
    .reply(200, { status: 'success', message: 'd0b2bd07-d49e-5ba1-9b05-ec23ac921930' })

    return uport.push(PUSHTOKEN, {url: payloadUrl}).then(response => {
      return expect(response).toEqual({ status: 'success', message: 'd0b2bd07-d49e-5ba1-9b05-ec23ac921930' })
    })
  })

  it('handles missing token', () => {
    return uport.push(null, {url: payloadUrl}).catch(error => expect(error.message).toEqual('Missing push notification token'))
  })

  it('handles missing payload', () => {
    return uport.push(PUSHTOKEN, {}).catch(error => expect(error.message).toEqual('Missing payload url for sending to users device'))
  })

  it('handles invalid token', () => {
    nock('https://pututu.uport.me', {
      reqheaders: {
        'authorization': `Bearer ${PUSHTOKEN}`
      }
    })
    .post('/api/v1/sns', {url: payloadUrl})
    .reply(403, 'Not allowed')

    return uport.push(PUSHTOKEN, {url: payloadUrl}).catch(error => expect(error.message).toEqual('Error sending push notification to user: Invalid Token'))
  })

  it('handles random error', () => {
    nock('https://pututu.uport.me', {
      reqheaders: {
        'authorization': `Bearer ${PUSHTOKEN}`
      }
    })
    .post('/api/v1/sns', {url: payloadUrl})
    .reply(500, 'Server Error')

    return uport.push(PUSHTOKEN, {url: payloadUrl}).catch(error => expect(error.message).toEqual('Error sending push notification to user: 500 Server Error'))
  })
})

describe('registry', () => {
  it('has a default registry that looks up profile', () => {
    return new Credentials().settings.registry('0x3b2631d8e15b145fd2bf99fc5f98346aecdc394c').then(profile =>
      expect(profile.publicKey).toEqual('0x0482780d59037778ea03c7d5169dd7cf47a835cb6d57a606b4e6cf98000a28d20d6d6bfae223cc76fd2f63d8a382a1c054788c4fafb1062ee89e718b96e0896d40')
    )
  })

  it('has ability to lookup profile', () => {
    return new Credentials().lookup('0x3b2631d8e15b145fd2bf99fc5f98346aecdc394c').then(profile =>
      expect(profile.publicKey).toEqual('0x0482780d59037778ea03c7d5169dd7cf47a835cb6d57a606b4e6cf98000a28d20d6d6bfae223cc76fd2f63d8a382a1c054788c4fafb1062ee89e718b96e0896d40')
    )
  })
})

describe('configNetworks', () => {
  it('should accept a valid network setting', () => {
    const networks = {'0x94365e3b': { rpcUrl: 'https://private.chain/rpc', registry: '0x3b2631d8e15b145fd2bf99fc5f98346aecdc394c' }}
    const credentials =  new Credentials({networks})
    expect(credentials.settings.networks).toEqual(networks)
  })

  it('should require a registry address', () => {
    const networks = {'0x94365e3b': { rpcUrl: 'https://private.chain/rpc' }}
    expect(() => new Credentials({networks})).toThrowErrorMatchingSnapshot()
  })

  it('should require a rpcUrl', () => {
    const networks = {'0x94365e3b': { registry: '0x3b2631d8e15b145fd2bf99fc5f98346aecdc394c' }}
    expect(() => new Credentials({networks})).toThrowErrorMatchingSnapshot()
  })

  it('if networks key is passed in it must contain configuration object', () => {
    const networks = {'0x94365e3b': 'hey'}
    expect(() => new Credentials({networks})).toThrowErrorMatchingSnapshot()
  })
})