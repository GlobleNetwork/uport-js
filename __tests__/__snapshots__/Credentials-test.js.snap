exports[`attest has correct payload in JWT for an attestation 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "claim": Object {
      "email": "bingbangbung@email.com",
    },
    "exp": 1485321134,
    "iat": 1485321133,
    "iss": "0x001122",
    "sub": "0x112233",
  },
  "signature": "kO5mGjkW6vkvE_JjWfKrt-ZSITo_EXoRIslhDCss78bM2458nQF-wBiFzIWtvjC38TR2NrLh6trz8wG_auTEig",
}
`;

exports[`configNetworks if networks key is passed in it must contain configuration object 1`] = `"Network configuration object required"`;

exports[`configNetworks should require a registry address 1`] = `"Malformed network config object, object must have \'registry\' key specified."`;

exports[`configNetworks should require a rpcUrl 1`] = `"Malformed network config object, object must have \'rpcUrl\' key specified."`;

exports[`createAttestationRequest has correct payload in JWT for a request 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1485321133,
    "iss": "0x001122",
    "sub": "did:uport:223ab45",
    "type": "attReq",
    "unsignedClaims": Array [
      Object {
        "claim": Object {
          "test": Object {
            "prop1": 1,
            "prop2": 2,
          },
        },
      },
    ],
  },
  "signature": "YM7DQRvh5vi4pLlWPJ5JHacmxJarnt2HhfYqwd0njEwwcYLsSMaK0yPFPDUqXwbdDdeLE7Sbc1ezbLm_Fw78TQ",
}
`;

exports[`createRequest has correct payload in JWT for a plain request for public details 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "type": "shareReq",
  },
  "signature": "kUs_SxnmUvmAa7-NKJ_qiIoODAB_MgCEMdmkw6pGy3ULMwQ-KWh6-M5CYkvPxPyHoDHVB5yGcFmTMxp_PU8a-g",
}
`;

exports[`createRequest has correct payload in JWT for a request 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "n8RNFvH4VY4BGM9Tvm9sw_UlPMi5dLX66bkCv8I_6TGcyS7trpPsjttZjilYnGCakzWtnok9D9aqFsd2W5Qkng",
}
`;

exports[`createRequest has correct payload in JWT for a request asking for verified credentials 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
    "verified": Array [
      "name",
    ],
  },
  "signature": "qmFvkl3wvtQiV7QAlLEljP-Uv_4wdNqsgcX_c9T0DFWfQE_p2nI0XugDHOAyF6_DTDhMxoNZvhLq6awUmlFGHA",
}
`;

exports[`createRequest has correct payload in JWT for a request for push notifications 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "permissions": Array [
      "notifications",
    ],
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "GNWPTAF9blrt7TtMps-j6Lt28bM9zzGCmj6S10nmLltgWIM4fLamibv2w1ktt-LtIFlU6cPTsnlOJr0bCi49CQ",
}
`;

exports[`createRequest has correct payload in JWT for a request with callbackUrl 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "callback": "https://myserver.com",
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "5OxjY-V3fdvRyv4XF5fayeQI9EiDZFrx47AUMnh8WnpUQes8qw6vs12bP9F0OnjEfYCkBmM-hTcq7OtTQBK4Ww",
}
`;

exports[`createRequest has correct payload in JWT requesting a bad accountType 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "type": "shareReq",
  },
  "signature": "kUs_SxnmUvmAa7-NKJ_qiIoODAB_MgCEMdmkw6pGy3ULMwQ-KWh6-M5CYkvPxPyHoDHVB5yGcFmTMxp_PU8a-g",
}
`;

exports[`createRequest has correct payload in JWT requesting a specific accountType 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "act": "devicekey",
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "type": "shareReq",
  },
  "signature": "eJaJnJwIXz91Yuo522likH5HX8DrLRoZGayAcvTPXBQ5FnZePjUAPmM03cklPyejdBfhUqIHezXwn146pvvJDA",
}
`;

exports[`createRequest has correct payload in JWT requesting a specific network_id 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "net": "0x4",
    "type": "shareReq",
  },
  "signature": "itf9cAdqqOz635UidRXvCIV6U3DNbzBfDGivELfjypXbRGZgTS7CyDSJxjKwSRxGRqyhof9XzEKqzrldngTeAA",
}
`;

exports[`createRequest ignores unsupported request parameters 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "exp": 1485321733,
    "iat": 1485321133,
    "iss": "0x001122",
    "type": "shareReq",
  },
  "signature": "kUs_SxnmUvmAa7-NKJ_qiIoODAB_MgCEMdmkw6pGy3ULMwQ-KWh6-M5CYkvPxPyHoDHVB5yGcFmTMxp_PU8a-g",
}
`;

exports[`receive returns profile mixing public and private claims 1`] = `
Object {
  "address": "0x001122",
  "name": "Davie",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile mixing public and private claims 2`] = `
Object {
  "address": "0x001122",
  "name": "Davie",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile mixing public and private claims and verified credentials 1`] = `
Object {
  "address": "0x001122",
  "name": "Davie",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
  "verified": Array [
    Object {
      "claim": Object {
        "email": "bingbangbung@email.com",
      },
      "exp": 1485321134,
      "iat": 1485321133,
      "iss": "0x001122",
      "jwt": "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJpc3MiOiIweDAwMTEyMiIsImlhdCI6MTQ4NTMyMTEzMywic3ViIjoiMHgxMTIyMzMiLCJjbGFpbSI6eyJlbWFpbCI6ImJpbmdiYW5nYnVuZ0BlbWFpbC5jb20ifSwiZXhwIjoxNDg1MzIxMTM0fQ.kO5mGjkW6vkvE_JjWfKrt-ZSITo_EXoRIslhDCss78bM2458nQF-wBiFzIWtvjC38TR2NrLh6trz8wG_auTEig",
      "sub": "0x112233",
    },
  ],
}
`;

exports[`receive returns profile mixing public and private claims and verified credentials 2`] = `
Object {
  "address": "0x001122",
  "name": "Davie",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
  "verified": Array [
    Object {
      "claim": Object {
        "email": "bingbangbung@email.com",
      },
      "exp": 1485321134,
      "iat": 1485321133,
      "iss": "0x001122",
      "jwt": "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJpc3MiOiIweDAwMTEyMiIsImlhdCI6MTQ4NTMyMTEzMywic3ViIjoiMHgxMTIyMzMiLCJjbGFpbSI6eyJlbWFpbCI6ImJpbmdiYW5nYnVuZ0BlbWFpbC5jb20ifSwiZXhwIjoxNDg1MzIxMTM0fQ.kO5mGjkW6vkvE_JjWfKrt-ZSITo_EXoRIslhDCss78bM2458nQF-wBiFzIWtvjC38TR2NrLh6trz8wG_auTEig",
      "sub": "0x112233",
    },
  ],
}
`;

exports[`receive returns profile with device key claims 1`] = `
Object {
  "address": "0x001122",
  "deviceKey": "0xdeviceKey",
  "name": "David Chaum",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile with only public claims 1`] = `
Object {
  "address": "0x001122",
  "name": "David Chaum",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile with only public claims 2`] = `
Object {
  "address": "0x001122",
  "name": "David Chaum",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile with private chain network id claims 1`] = `
Object {
  "address": "0x001122",
  "name": "David Chaum",
  "networkAddress": "34wjsxwvduano7NFC8ujNJnFjbacgYeWA8m",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile with private chain network id claims 2`] = `
Object {
  "address": "0x001122",
  "name": "David Chaum",
  "networkAddress": "34wjsxwvduano7NFC8ujNJnFjbacgYeWA8m",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;
